### PR TITLE
Deprecate LightningModule.model_size

### PR DESCRIPTION
**Pull request by roshikouhai**
## What does this PR do?

This PR deprecates the model_size property in LightningModule as discussed and agreed in #8343 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]
Thanks for working on this @roshikouhai !

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1964,8 +1964,14 @@ def model_size(self) -> float:
...
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0
         """
-        # todo: think about better way without need to dump model to drive
+        rank_zero_deprecation(
+            'The `LightningModule.model_size` property was deprecated in v1.4 and will be removed in v1.6.'
+        )
+
         tmp_name = f"{uuid.uuid4().hex}.pt"
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1975]: Should we move this a utility function instead ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1964,8 +1964,14 @@ def model_size(self) -> float:
         """
         The model's size in megabytes. The computation includes everything in the
         :meth:`~torch.nn.Module.state_dict`, i.e., by default the parameteters and buffers.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0
         """
-        # todo: think about better way without need to dump model to drive
+        rank_zero_deprecation(
+            'The `LightningModule.model_size` property was deprecated in v1.4 and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1972]: I miss what is the alternative here?

**Review by Borda — CHANGES_REQUESTED** [review]
if so, pls replace all usage in the codebase...

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (roshikouhai) [reply pytorch_lightning/core/lightning.py:1975]: Should I move it to a utility function in this PR or create a new PR?

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (roshikouhai) [reply pytorch_lightning/core/lightning.py:1972]: @Borda the alternative would be to move this function to an utility function, I am just not sure to include that within this PR or a new one.

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1975]: I think it would be reasonable to do it in this PR because we would have to update the deprecation message anyway (saying that it has been moved to ...)

Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1975]: there is a parallel PR for this in https://github.com/PyTorchLightning/pytorch-lightning/pull/8500 - should we merge that and then update the message here or deprecation?

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1972]: @roshikouhai we can pull the changes from https://github.com/PyTorchLightning/pytorch-lightning/pull/8500/ into this PR to make this deprecation clearer. 

We can also update existing callers of this property to the new utility

**Comment by roshikouhai** [comment]
Rebased on top of #8500. I tried updating [test_quantization.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/callbacks/test_quantization.py) to use the new get_model_size() helper function but as @calebrob6 states, it currently doesn't work for models with sparse tensors which the tests have. Not sure if I should leave it be or find a way to implement the sparse tensors in this PR.

**Comment by calebrob6** [comment]
@roshikouhai, an alternate way to do `get_model_size(...)` is:

```
def get_model_size(self, model):
    with io.BytesIO() as f:
        torch.save(model.state_dict(), f)
        f.seek(0, io.SEEK_END)
        size_mb = f.tell() / 1e6
    return size_mb
 ```

I preferred to _not_ do it this way as this will use quite a bit of RAM. Would you like me to look into how to implement getting the size of the sparse tensors as well?

**Comment by ananthsub** [comment]
@roshikouhai @calebrob6 - as the primary motivation for the PR is to remove this function off the core LightningModule API, what do you think about this sequencing?
1. Define the utility function which re-uses the exact same implementation and deprecate this off the core API and point users to the new utility
2. Later (e.g. for v1.5) investigate improvements we can make to implementation of the new utility to avoid file saving while supporting sparse tensors

**Comment by roshikouhai** [comment]
> Later (e.g. for v1.5) investigate improvements we can make to implementation of the new utility to avoid file saving while supporting sparse tensors 

Given the goal of this PR is to deprecate model_size from LightningModule, I vote towards this option and allow @calebrob6  the time he needs to implement the model_size for sparse tensors.

File: tests/utilities/test_model_helpers.py [hunk 3]
```diff
@@ -92,3 +94,24 @@ def on_fit_start(self):
     trainer = Trainer(fast_dev_run=1)
     trainer.fit(model, train_dataloader=model.train_dataloader())
     assert model.on_fit_start_called
+
+
+def test_get_model_size():
+    model = BoringModel()
+
+    size_bytes = get_model_size(model)
+    assert size_bytes == 0.001319
```
Comment (roshikouhai) [inline tests/utilities/test_model_helpers.py:103]: As we want to keep the original implementation of model_size, changed @calebrob6's implementation from using bytes to megabytes.

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1966,8 +1966,15 @@ def model_size(self) -> float:
...
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0
         """
-        # todo: think about better way without need to dump model to drive
+        rank_zero_deprecation(
+            'The `LightningModule.model_size` property was deprecated in v1.4 and will be removed in v1.6. Please '
+            'use the get_model_size method under utilities/model_helpers.py'
+        )
+
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1977]: Now that 1.4 is released, this will be officially marked as deprecated in v1.5

File: pytorch_lightning/utilities/model_helpers.py [hunk 5]
```diff
@@ -69,3 +74,19 @@ def is_overridden(
...
+
+    Returns:
+        Number of megabytes in the parameters of the input module
+    """
+    # TODO: Implement a method without needing to download the model
+    tmp_name = f"{uuid.uuid4().hex}.pt"
+    torch.save(model.state_dict(), tmp_name)
+    size_mb = os.path.getsize(tmp_name) / 1e6
+    os.remove(tmp_name)
+    return size_mb
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_helpers.py:92]: IMO this makes more sense here instead of `model_helpers.py`: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/memory.py

File: pytorch_lightning/utilities/model_helpers.py [hunk 6]
```diff
@@ -69,3 +74,19 @@ def is_overridden(
     parent_code = parent_attr.__code__
 
     return instance_code != parent_code
+
+
+def get_model_size(model: nn.Module) -> int:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_helpers.py:79]: let's include the units in the name so readers of the code immediately know that the returned value is in MB
```suggestion
def get_model_size_mb(model: nn.Module) -> int:
```

File: pytorch_lightning/utilities/model_helpers.py [hunk 7]
```diff
@@ -69,3 +74,19 @@ def is_overridden(
...
+    """
+    Calculates the size of a nn.Module in bytes by saving the model to a temprorary file
+    and reading in the size.
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_helpers.py:82]:
```suggestion
    Calculates the size of a nn.Module in megabytes by saving the model to a temporary file
    and reading in the size.
```

File: tests/deprecated_api/test_remove_1-6.py [hunk 8]
```diff
@@ -330,3 +330,12 @@ def test_v1_6_0_deprecated_device_dtype_mixin_import():
     _soft_unimport_module('pytorch_lightning.utilities.device_dtype_mixin')
     with pytest.deprecated_call(match='will be removed in v1.6'):
         from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin  # noqa: F811 F401
+
+
+def test_v1_6_0_deprecated_model_size():
```
Comment (ananthsub) [inline tests/deprecated_api/test_remove_1-6.py:335]: this will be in the test_remove_1-7.py now

File: tests/utilities/test_memory.py [hunk 9]
```diff
@@ -28,3 +30,24 @@ def test_recursive_detach():
...
+def test_get_sparse_model_size_mb():
+    class BoringSparseModel(BoringModel):
+        def __init__(self):
+            super().__init__()
+            self.layer = nn.Parameter(torch.ones(32).to_sparse())
+
+    model = BoringSparseModel()
+    size_bytes = get_model_size_mb(model)
+
+    assert size_bytes == 0.001511 or size_bytes == 0.001665
```
Comment (ananthsub) [inline tests/utilities/test_memory.py:53]: you can use `math.isclose()` for the floating point comparisons in case of flakiness

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
@Borda do you see any remaining issues? is this good to go?

File: pytorch_lightning/core/lightning.py [hunk 10]
```diff
@@ -1959,7 +1959,11 @@ def model_size(self) -> float:
         The model's size in megabytes. The computation includes everything in the
         :meth:`~torch.nn.Module.state_dict`, i.e., by default the parameteters and buffers.
         """
-        # todo: think about better way without need to dump model to drive
+        rank_zero_deprecation(
+            "The `LightningModule.model_size` property was deprecated in v1.5 and will be removed in v1.7. Please "
+            "use the get_model_size_mb method under utilities/memory.py"
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1964]:
```suggestion
            "use the `utilities.memory.get_model_size_mb` method under utilities/memory.py"
```

**Review by Borda — APPROVED** [review]
fixed the remaining

File: tests/deprecated_api/test_remove_1-7.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import pytest
+
+from tests.helpers import BoringModel
+
+
+def test_v1_7_0_deprecated_model_size():
+    model = BoringModel()
+    with pytest.deprecated_call(
+        match="The `LightningModule.model_size` property was deprecated in v1.5 and will be removed in v1.7. "
+        "Please use the get_model_size_mb method under utilities/memory.py"
```
Comment (Borda) [inline tests/deprecated_api/test_remove_1-7.py:24]:
```suggestion
        "Please use the `utilities.memory.get_model_size_mb` method under utilities/memory.py"
```

**Review by carmocca — APPROVED** [review]
Pushed a commit to call the new function from the old one
